src/Microsoft.Azure.SignalR.Common/Interfaces/AsyncLocalScopeBase.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.Azure.SignalR
+{
+    // todo [wanl]: protect asyncLocal, prevent from anyone can access to async local
+    internal abstract class AsyncLocalScopeBase<TProps> : IDisposable
+    {
+        public static bool IsScopeEstablished => _asyncLocal.Value != null;
+
+        private bool _needCleanup;
+
+        private static AsyncLocal<Context> _asyncLocal = new AsyncLocal<Context>();
+
+        public AsyncLocalScopeBase(TProps props)
+        {
+            if (!IsScopeEstablished)
+            {
+                _needCleanup = true;
+                _asyncLocal.Value = new Context { Value = new Value { Properties = props } };
+            }
+            else
+            {
+                Debug.Assert(!IsScopeEstablished, "Attempt to replace an already established scope");
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_needCleanup)
+            {
+                // shallow cleanup since we don't want any execution contexts in unawaited tasks 
+                // to suddenly change behavior once we're done with disposing
+                _asyncLocal.Value = null;
+            }
+        }
+
+        protected static Context Current
+        {
+            get => _asyncLocal.Value;
+            set => _asyncLocal.Value = value;
+        }
+
+        protected class Context

```

Comment:
oh no, what are these nested classes...?

Reply:
remember in this [pr](https://github.com/Azure/azure-signalr/pull/912)? 
Context -> [ClientConnectionScopeProperties](https://github.com/Azure/azure-signalr/blob/a409e1a3f30f43e1bc35318c70b212a61e329b71/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ClientConnectionScope.cs#L95)
Value -> [DiagnosticLogsContext](https://github.com/Azure/azure-signalr/blob/a409e1a3f30f43e1bc35318c70b212a61e329b71/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ClientConnectionScope.cs#L101)
We need to add a reference to make that we are able to update `ServiceDiagnosticLogsContext` in event handler.

Reply:
seperated